odl/discr/partition.py

```diff
@@ -546,6 +555,139 @@ def squeeze(self):
         newset = self.set[self.grid.inondeg]
         return RectPartition(newset, self.grid.squeeze())
 
+    def index(self, value, floating=False):
+        """Return the index of a value in the domain.
+
+        Parameters
+        ----------
+        value : ``self.set`` element
+            The point to find the index of.
+        floating : bool
+            True if the index should also give the position inside the voxel.
+            This is given by returning the integer valued index of the voxel
+            plus the partial distance from the left to the right boundary.
+
+        Returns
+        -------
+        index : int/float or tuple of int/float
+            Index of the value, as counted from the left.
+            If ``self.ndim>1`` the result is a tuple, else a scalar.
+            If ``floating=True`` the scalar is a float, else an int.
+
+        Examples
+        --------
+        Get the indices of start and end:
+
+        >>> p = uniform_partition(0, 2, 5)
+        >>> p.index(0)
+        0
+        >>> p.index(2)
+        4
+
+        Also supports points inside voxels, in that case returning the index
+        of the voxel:
+
+        >>> p.index(0.2)
+        0
+
+        By using the ``floating`` argument, partial positions inside the voxels
+        can instead be determined:
+
+        >>> p.index(0.2, floating=True)
+        0.5
+
+        These indexes work with indexing, extracting the voxel the point lies
+        in:
+
+        >>> p[p.index(0.1)]
+        uniform_partition(0.0, 0.4, 1)
+
+        Index also works in higher dimensions:
+
+        >>> p = uniform_partition([0, -1], [1, 2], (4, 1))
+        >>> p.index([0.5, 2])
+        (2, 0)
+        >>> p[p.index([0.5, 2])]
+        uniform_partition([0.5, -1.0], [0.75, 2.0], [1, 1])
+        """
+        value = np.atleast_1d(self.set.element(value))
+        result = ()
+        for val, cell_bdry_vec in zip(value, self.cell_boundary_vecs):
+            ind = np.searchsorted(cell_bdry_vec, val)
+            if floating:
+                if cell_bdry_vec[ind] == val:
+                    # Value is on top of edge
+                    result += (float(ind),)
+                else:
+                    # interpolate between
+                    csize = float(cell_bdry_vec[ind] - cell_bdry_vec[ind - 1])
+                    result += (ind - (cell_bdry_vec[ind] - val) / csize,)
+            else:
+                if cell_bdry_vec[ind] == val and ind != len(cell_bdry_vec) - 1:
+                    # Value is on top of edge, but not last edge
+                    result += (ind,)
+                else:
+                    result += (ind - 1,)
+
+        if self.ndim == 1:
+            result = result[0]
+
+        return result
+
+    @property
+    def bydimension(self):
```

Comment:
You could also use a name like `as_outer` since you represent the partition as an outer product.